test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/ApiBehaviorApplicationModelProviderTest.cs

```diff
@@ -383,6 +384,31 @@ public void InferBindingSourceForParameter_ReturnsBodyForComplexTypes()
             Assert.Same(BindingSource.Body, result);
         }
 
+        [Fact]
+        public void InferParameterBindingSources_SetsCorrectBindingSourceForComplexTypesWithCancellationToken()
```

Comment:
Sorry, missed the early conversation - were we just missing tests for this scenario?

Reply:
No, if you take a look at the CI, you can see that this is a failing test. I removed the actual fix because of conflicting changes with @dougbu. I'm not really sure where to go from here. The main culprit seems to be that `parameter.BindingInfo?.BindingSource` ends up being `null` here:

https://github.com/aspnet/Mvc/blob/fa629cd5e112d89db7cef7374694eb08dbc6bf9c/src/Microsoft.AspNetCore.Mvc.Core/Internal/ApiBehaviorApplicationModelProvider.cs#L154-L158

Reply:
Looks like `ModelMetadataProvider` isn't used by the default provider to set up the binding source. I'll have to consult with @dougbu \ @rynowak to figure out if this is something we could calculate as part of creating the `ParameterModel`.

Reply:
Looks like we've got two issues here:
- `DefaultApplicationModelProvider` calls `BindingInfo.GetBindingInfo(...)` to guess at the `ParameterModelBase.BindingInfo` rather than using an `IModelMetadataProvider`.
- Later, `PageBinderFactory` and similar `ParameterBinder` delegate creators ignore `ParameterDescriptor.BindingInfo`, likely knowing it was a guess copied from `ParameterModelBase.BindingInfo`.

The inconsistency means `DefaultApplicationModelProvider`, `IApplicationModelProvider` instances that run after it, and all `ParameterDescriptor` consumers receive guesswork.

Not sure there's a downside to `DefaultApplicationModelProvider` relying on an `IModelMetadataProvider` to determine the `BindingInfo`. If we also change the `ParameterBinder` delegate creators to use the `BindingInfo` they're given, changes made to this information in `IApplicationModelProvider` will actually matter when binding 😃 Of course, we might need a back-compatibility conversation&hellip;